### PR TITLE
support aggregated access control definitions

### PR DESCRIPTION
**Pull request by tomaszpowroznik**
`[
  {
    "LiveFrom": "01/01/2017",
    "CaseTypeID": "CARE_SUPERVISION_EPO",
    "CaseEventID": "openCase",
    "UserRole": "caseworker-publiclaw-solicitor",
    "CRUD": "CRU"
  },
  {
    "LiveFrom": "01/01/2017",
    "CaseTypeID": "CARE_SUPERVISION_EPO",
    "CaseEventID": "deleteApplication",
    "AccessControl": [
      {
        "UserRoles": ["caseworker-publiclaw-solicitor", "caseworker-publiclaw-courtadmin"],
        "CRUD": "CRU"
      },
      {
        "UserRoles": ["caseworker-publiclaw-gatekeeper"],
        "CRUD": "R"
      }
    ]
  },
  {
    "LiveFrom": "01/01/2017",
    "CaseTypeID": "CARE_SUPERVISION_EPO",
    "CaseEventID": "ordersNeeded",
    "UserRoles": ["caseworker-publiclaw-solicitor", "caseworker-publiclaw-courtadmin"],
    "CRUD": "CRU"
  },
  {
    "LiveFrom": "01/01/2017",
    "CaseTypeID": "CARE_SUPERVISION_EPO",
    "CaseEventID": "ordersNeeded",
    "UserRoles": ["caseworker-publiclaw-gatekeeper", "caseworker-publiclaw-judiciary"],
    "CRUD": "D"
  }
]
`

### CODE REVIEW

File: src/main/lib/extensions.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const ACCESS_CONTROL = 'AccessControl';
+const USER_ROLES = 'UserRoles';
+const USER_ROLE = 'UserRole';
+const CRUD = 'CRUD';
+
+const expandRoles = json => {
+  const newJson = new Array();
```
Comment (lewisbirks) [inline src/main/lib/extensions.js:7]: Can just be const newJson = [];

**Review by lewisbirks — CHANGES_REQUESTED** [review]
Looks good, can we get some tests for the new function

File: src/main/lib/extensions.js (hunk 1 above)
Comment (tomaszpowroznik) [reply src/main/lib/extensions.js]: done

File: src/test/access-control-transformer.test.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
+const assert = require('assert');
+const accessControlTransformer = require('../main/lib/access-control-transformer');
+
+describe('Access control transformer', () => {
+
+  it('should do nothing if AccessControl and UserRoles are not present', async () => {
```
Comment (lewisbirks) [inline src/test/access-control-transformer.test.js:6]: Is there a reason that these tests are asynchronous?

File: src/test/access-control-transformer.test.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
+const assert = require('assert');
```
Comment (lewisbirks) [inline src/test/access-control-transformer.test.js:1]: Was just looking through the assert docs, seems like it advises the usage of `require('assert').strict` over without it.

File: src/main/lib/access-control-transformer.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        if (!lodash.isArray(usersRoles) || lodash.isEmpty(usersRoles)) {
+          throw new Error(`${ACCESS_CONTROL} requires non empty ${USER_ROLES} array.`);
+        }
+
+        if (lodash.isEmpty(crud)) {
+          throw new Error(`${ACCESS_CONTROL} requires non empty ${CRUD} field.`);
+        }
+
+        usersRoles.forEach(userRole => newJson.push(lodash(obj).omit(ACCESS_CONTROL).assign({[USER_ROLE]: userRole, [CRUD]: crud}).value()));
```
Comment (lewisbirks) [inline src/main/lib/access-control-transformer.js:40]: Do you think this more readable?
```suggestion
        usersRoles.forEach(userRole => {
          newJson.push(lodash(obj).omit(ACCESS_CONTROL).assign({
            [USER_ROLE]: userRole,
            [CRUD]: crud
          }).value());
        });
```

File: src/main/lib/access-control-transformer.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        throw new Error(`${USER_ROLES} must be non empty array.`);
+      }
+
+      if (obj[USER_ROLE]) {
+        throw new Error(`${USER_ROLES} and ${USER_ROLE} not allowed on the same element.`);
+      }
+
+      const crud = obj[CRUD];
+
+      usersRoles.forEach(userRole => newJson.push(lodash(obj).omit(USER_ROLES).assign({[USER_ROLE]: userRole, [CRUD]: crud}).value()));
```
Comment (lewisbirks) [inline src/main/lib/access-control-transformer.js:53]: Do you think it is more readable having it split into multiple lines
```suggestion
      usersRoles.forEach(userRole => {
        newJson.push(lodash(obj).omit(USER_ROLES).assign({
          [USER_ROLE]: userRole,
          [CRUD]: crud
        }).value());
      });
```

File: src/main/lib/access-control-transformer.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+const lodash = require('lodash');
```
Comment (lewisbirks) [inline src/main/lib/access-control-transformer.js:1]: Do you need to add lodash as a dependancy in the package.json?

**Review by lewisbirks — COMMENTED** [review]
Looks good, a few comments

File: src/test/access-control-transformer.test.js (hunk 2 above)
Comment (tomaszpowroznik) [reply src/test/access-control-transformer.test.js]: the only reason is that i'm lazy and did copy paste, of course not needed. Done

File: src/test/access-control-transformer.test.js (hunk 3 above)
Comment (tomaszpowroznik) [reply src/test/access-control-transformer.test.js]: done

File: src/main/lib/access-control-transformer.js (hunk 4 above)
Comment (tomaszpowroznik) [reply src/main/lib/access-control-transformer.js]: agree

File: src/main/lib/access-control-transformer.js (hunk 5 above)
Comment (tomaszpowroznik) [reply src/main/lib/access-control-transformer.js]: done

File: src/test/access-control-transformer.test.js (hunk 2 above)
Comment (lewisbirks) [reply src/test/access-control-transformer.test.js]: The best reason

File: src/main/lib/access-control-transformer.js (hunk 6 above)
Comment (tomaszpowroznik) [reply src/main/lib/access-control-transformer.js:1]: thats good spot, i have unconsciously piggyback on transitive dependency

**Review by lewisbirks — APPROVED** [review]
LGTM

**Pull request merged**